ksqldb-rest-app/src/test/java/io/confluent/ksql/api/integration/ApiIntegrationTest.java

```diff
@@ -659,6 +678,233 @@ private void shouldFailToExecuteQuery(final String sql, final String message) {
+  }
+
+  @Test
+  public void shouldExecutePrintQueryFromBeginningWithLimit() {
+    // Given:
+    String sql = "PRINT " + TEST_TOPIC + " FROM BEGINNING LIMIT 3;";
+
+    // When:
+    AtomicReference<PrintResponse> atomicReference = new AtomicReference<>();
+    assertThatEventually(() -> {
+      PrintResponse printResponse = executePrintTopic(sql);
+      atomicReference.set(printResponse);
+      return printResponse.rows.size();
+    }, is(3));
+
+    PrintResponse printResponse = atomicReference.get();
+
+    // Then:
+    assertThat(printResponse.rows.get(0), containsString("rowtime:"));
+    assertThat(printResponse.rows.get(0), containsString(
+        "key: {\"F1\":[\"a\"]}, value: {\"STR\":\"FOO\",\"LONG\":1,\"DEC\":1.11,\"BYTES_\":\"AQ==\",\"ARRAY\":[\"a\"],\"MAP\":{\"k1\":\"v1\"},\"STRUCT\":{\"F1\":2},\"COMPLEX\":{\"DECIMAL\":0.0,\"STRUCT\":{\"F1\":\"v0\",\"F2\":0},\"ARRAY_ARRAY\":[[\"foo\"]],\"ARRAY_STRUCT\":[{\"F1\":\"v0\"}],\"ARRAY_MAP\":[{\"k1\":0}],\"MAP_ARRAY\":{\"k\":[\"v0\"]},\"MAP_MAP\":{\"k\":{\"k\":0}},\"MAP_STRUCT\":{\"k\":{\"F1\":\"v0\"}}},\"TIMESTAMP\":1,\"DATE\":1,\"TIME\":0}, partition: 0"));
+    assertThat(printResponse.rows.get(1), containsString("rowtime:"));
+    assertThat(printResponse.rows.get(1), containsString(
+        "key: {\"F1\":[\"b\"]}, value: {\"STR\":\"BAR\",\"LONG\":2,\"DEC\":2.22,\"BYTES_\":\"Ag==\",\"ARRAY\":[],\"MAP\":{},\"STRUCT\":{\"F1\":3},\"COMPLEX\":{\"DECIMAL\":1.0,\"STRUCT\":{\"F1\":\"v1\",\"F2\":1},\"ARRAY_ARRAY\":[[\"foo\"]],\"ARRAY_STRUCT\":[{\"F1\":\"v1\"}],\"ARRAY_MAP\":[{\"k1\":1}],\"MAP_ARRAY\":{\"k\":[\"v1\"]},\"MAP_MAP\":{\"k\":{\"k\":1}},\"MAP_STRUCT\":{\"k\":{\"F1\":\"v1\"}}},\"TIMESTAMP\":2,\"DATE\":2,\"TIME\":1}, partition: 0"));
+    assertThat(printResponse.rows.get(2), containsString("rowtime:"));
+    assertThat(printResponse.rows.get(2), containsString(
+        "key: {\"F1\":[\"c\"]}, value: {\"STR\":\"BAZ\",\"LONG\":3,\"DEC\":30.33,\"BYTES_\":\"Aw==\",\"ARRAY\":[\"b\"],\"MAP\":{},\"STRUCT\":{\"F1\":null},\"COMPLEX\":{\"DECIMAL\":2.0,\"STRUCT\":{\"F1\":\"v2\",\"F2\":2},\"ARRAY_ARRAY\":[[\"foo\"]],\"ARRAY_STRUCT\":[{\"F1\":\"v2\"}],\"ARRAY_MAP\":[{\"k1\":2}],\"MAP_ARRAY\":{\"k\":[\"v2\"]},\"MAP_MAP\":{\"k\":{\"k\":2}},\"MAP_STRUCT\":{\"k\":{\"F1\":\"v2\"}}},\"TIMESTAMP\":3,\"DATE\":3,\"TIME\":2}, partition: 0"));
+  }
+
+  @Test
+  public void shouldExecutePrintQueryFromEarliestOffsetWithLimit() {
+    // Given:
+    String sql = "PRINT " + TEST_TOPIC + " LIMIT 3;";
+
+    // When:
+    AtomicReference<PrintResponse> atomicReference = new AtomicReference<>();
+    assertThatEventually(() -> {
+      JsonObject printProperties = new JsonObject().put("auto.offset.reset", "earliest");
```

Comment:
Is it true that `auto.offset.reset` always takes priority? For example, if `auto.offset.reset` is explicitly set to latest and `PRINT <topic> FROM BEGINNING;` is issued, does that mean the print statement executes from latest rather than from beginning? It'd be good to double check that whatever behavior `/query-stream` produces is consistent with the existing handling for PRINT statements. (I assume it is but never hurts to check.)

Reply:
I made it the other way around, which I think is what you meant in the last review and what I also think is correct. The statement has priority over the config. That means, `PRINT <topic> FROM BEGINNING` issued with `auto.offset.reset = latest` will print from the beginning. https://github.com/confluentinc/ksql/pull/9721#issuecomment-1339695003

For `/query-stream` we only have `"auto.offset.reset"`, we can't set the cursor position in the SQL Statement, right? That's why I think what we have is consistent. Since the PRINT statement has a specific clause for this, this has priority. If not explicity when sending the PRINT SQL, we fall back to the default behaviour for the `/query-stream` endpoint: using whatever is set for `auto.offset.reset`.

Reply:
OK, IIUC the order of precedence is:
* if the statement contains FROM BEGINNING, then the topic is printed from the beginning (regardless of auto.offset.reset)
* otherwise, if the user explicitly sets `auto.offset.reset=earliest`, then the topic is printed from the beginning. If the user sets `auto.offset.reset=latest` or does not explicitly set a value for `auto.offset.reset`, then the topic is printed from the end.

Does that sound right? That appears consistent with what we do from the CLI so that sounds great if so.

Reply:
> Does that sound right? That appears consistent with what we do from the CLI so that sounds great if so.
Exactly! That sounds right :)

Reply:
Great! Thanks for following up on all of these and helping me understand :)

Reply:
Thank you for the review! Glad that we were able to get this done and merged :)